File: java/src/main/java/com/alibaba/graphar/types/FileType.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.alibaba.graphar.types;

import static com.alibaba.graphar.util.CppClassName.GAR_FILE_TYPE;

import com.alibaba.fastffi.CXXEnum;
import com.alibaba.fastffi.FFITypeAlias;
import com.alibaba.fastffi.FFITypeRefiner;

@FFITypeAlias(GAR_FILE_TYPE)
@FFITypeRefiner("com.alibaba.graphar.types.FileType.get")
public enum FileType implements CXXEnum {
    CSV,
    PARQUET,
    ORC;

    public static FileType get(int value) {
        switch (value) {
            case 0:
                return CSV;
            case 1:
                return PARQUET;
            case 2:
                return ORC;
            default:
                throw new IllegalStateException("Unknown value for file type: " + value);
        }
    }

    @Override
    public int getValue() {
        return ordinal();
    }
}

File: go/graphar/types/filetype.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package types

import "fmt"

// FileType is the on-disk physical format of a property group or adjacency
// list: csv / parquet / orc / json. (json parses but is rejected at validation
// time; see FileTypeJSON.)
type FileType uint8

// FileType values. Zero value is unset; constructed values are non-zero so
// callers can detect a forgotten assignment via IsZero().
const (
	// fileTypeUnset is the zero / unset sentinel. Not exported because it
	// is never a valid on-disk format.
	fileTypeUnset FileType = iota
	// FileTypeCSV is the comma-separated values text format.
	FileTypeCSV
	// FileTypeParquet is the Apache Parquet columnar format.
	FileTypeParquet
	// FileTypeORC is the Apache ORC columnar format.
	FileTypeORC
	// FileTypeJSON is the JSON text format. Parseable for read-only fixture
	// compatibility but rejected by PropertyGroup.Validate.
	FileTypeJSON
)

// IsZero reports whether f is the zero (unset) FileType.
func (f FileType) IsZero() bool { return f == fileTypeUnset }

// String returns the on-disk spelling. The unset zero value stringifies to
// "unset" (not a valid file type).
func (f FileType) String() string {
	switch f {
	case fileTypeUnset:
		return "unset"
	case FileTypeCSV:
		return "csv"
	case FileTypeParquet:
		return "parquet"
	case FileTypeORC:
		return "orc"
	case FileTypeJSON:
		return "json"
	default:
		return "unknown"
	}
}

// ParseFileType parses the on-disk spelling produced by FileType.String.
func ParseFileType(s string) (FileType, error) {
	switch s {
	case "csv":
		return FileTypeCSV, nil
	case "parquet":
		return FileTypeParquet, nil
	case "orc":
		return FileTypeORC, nil
	case "json":
		return FileTypeJSON, nil
	default:
		return fileTypeUnset, fmt.Errorf("%w: %q", ErrInvalidFileType, s)
	}
}
